app/src/app/views/stix/collection/collection-import/collection-import-workflow/collection-import.component.ts

```diff
@@ -19,266 +22,544 @@ import { Tactic } from 'src/app/classes/stix/tactic';
-                            next: (results) => { 
-                                if (results.import_categories.errors.length > 0) {
-                                    logger.warn("Collection import completed with errors:", results.import_categories.errors);
-                                }
-                                this.save_errors = results.import_categories.errors;
-                                let save_error_ids = new Set(this.save_errors.map(err => err['object_ref']));
-                                for (let category in results.import_categories) {
-                                    if (category == "errors") continue;
-                                    for (let id of results.import_categories[category]) if (!save_error_ids.has(id)) this.successfully_saved.add(id);
-                                }
-                                logger.log("Successfully imported the following objects:", Array.from(this.successfully_saved));
-                                this.stepper.next(); 
-                            },
-                            complete: () => { subscription.unsubscribe(); } //prevent memory leaks
-                        })
-                    })
-                }
-            },
-            complete: () => { promptSubscription.unsubscribe() } //prevent memory leaks
-        })
+    for (let object of collectionBundle.objects) {
+      // look up the category for the object
+      if (!(object.id in idToCategory)) {
+        // does not belong to a change category
+        this.unchanged_ids.push(object.id);
+        continue;
+      }
+      // track that this object has changed
+      this.changed_ids.push(object.id);
+      // determine the change category
+      let category = idToCategory[object.id];
+      // wrap the object as if it came from the back-end
+      let raw: { [key: string]: any } = { stix: object, workspace: {} };
+      // parse the object & add it to the appropriate category for rendering
+      switch (object.type) {
+        case 'attack-pattern': //technique
+          this.object_import_categories.technique[category].push(
+            new Technique(raw)
+          );
+          break;
+        case 'x-mitre-tactic': //tactic
+          this.object_import_categories.tactic[category].push(new Tactic(raw));
+          break;
+        case 'malware': //software
+        case 'tool':
+          this.object_import_categories.software[category].push(
+            new Software(object.type, raw)
+          );
+          break;
+        case 'relationship': //relationship
+          // build source and target objects if the source/target objects aren't being uploaded simultaneously (in case where only relationships are uploaded)
+          if (object.source_ref in idToSdo) {
+            raw.source_object = idToSdo[object.source_ref];
+          } else {
+            raw.source_object = {
+              stix: {
+                attackID: object.source_id,
```

Comment:
The source object of a relationship may not always have an ATT&CK ID, namely in the case of Data Component objects which don't support the ATT&CK ID field. I'd recommend adding a ternary operator here to check that the `source_id` exists: `attackID: object.source_id ? object.source_id : '';`. This will prevent WB from logging an erroneous error that the field is invalid in cases where the source object doesn't support ATT&CK IDs.